zebrad/src/components/mempool/crawler.rs

```diff
@@ -0,0 +1,99 @@
+//! Zebra Mempool crawler.
+//!
+//! The crawler periodically requests transactions from peers in order to populate the mempool.
+
+use std::time::Duration;
+
+use futures::{stream, StreamExt, TryStreamExt};
+use tokio::{sync::Mutex, time::sleep};
+use tower::{timeout::Timeout, BoxError, Service, ServiceExt};
+
+use zebra_network::{Request, Response};
+
+#[cfg(test)]
+mod tests;
+
+/// The number of peers to request transactions from per crawl event.
+const FANOUT: usize = 4;
+
+/// The delay between crawl events.
+const RATE_LIMIT_DELAY: Duration = Duration::from_secs(75);
+
+/// The time to wait for a peer response.
+///
+/// # Correctness
+///
+/// If this timeout is removed or set too high, the crawler may hang waiting for a peer to respond.
+///
+/// If this timeout is set too low, the crawler may fail to populate the mempool.
+const PEER_RESPONSE_TIMEOUT: Duration = Duration::from_secs(6);
+
+/// The mempool transaction crawler.
+pub struct Crawler<S> {
+    peer_set: Mutex<Timeout<S>>,
+}
+
+impl<S> Crawler<S>
+where
+    S: Service<Request, Response = Response, Error = BoxError> + Clone + Send + 'static,
+    S::Future: Send,
+{
+    /// Spawn an asynchronous task to run the mempool crawler.
+    pub fn spawn(peer_set: S) {
+        let crawler = Crawler {
+            peer_set: Mutex::new(Timeout::new(peer_set, PEER_RESPONSE_TIMEOUT)),
+        };
+
+        tokio::spawn(crawler.run());
```

Comment:
This code drops the join guard, which means that panics and errors in the `Crawler` don't terminate the main thread.

We might want to do something similar to `poll_background_errors`:
https://github.com/ZcashFoundation/zebra/blob/main/zebra-network/src/peer_set/set.rs#L155

But that can wait for a follow-up ticket in the next sprint.
(We'll need some way of propagating errors from all the mempool tasks and futures.)

Reply:
It seemed like a simple fix, so I gave it a shot ([1d743d1](https://github.com/jvff/zebra/commit/1d743d14b85d57a505ceef0a9ed19d3366dccba6) and [a550557](https://github.com/jvff/zebra/commit/a5505575cb7ad1e54d75045241136afa116ea5a2)).

Reply:
Looks good!

We might also want the task to exit if the channel has been closed.
Can you add a note to the relevant channel ticket?

Reply:
Cool, I added it to #2603. Let me know if it's correct :sweat_smile: